### PR TITLE
Tensorflow object detection api fixes #1964 

### LINKED ISSUE #1964: [BUG] mlflow.tensorflow.autolog() does not work with tensorflow object detection api.
### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Linux Ubuntu 18.04
- **MLflow installed from (source or binary)**:binary
- **MLflow version (run ``mlflow --version``)**:1.3.0
- **Python version**:3.6.5
- **npm version, if running the dev UI**:

### Describe the problem
I am trying to use the **mlflow.tensorflow.autolog()** with [tensorflow object detection api](https://github.com/tensorflow/models/tree/master/research/object_detection). Adding **mlflow.tensorflow.autolog()** to **[model_main.py](https://github.com/tensorflow/models/blob/master/research/object_detection/model_main.py)** logs some parameters like global_norm/clipped_gradient_norm,global_norm/gradient_norm,global_step/sec,learning_rate_1,loss_1,loss_2 in mlflow. However the more important metrics linke map,precision,recall are not being logged in mlfow even if they are present in tensorboard.I used the example [here](https://github.com/mlflow/mlflow/blob/master/examples/tensorflow/train_predict.py) as reference but did

### PR DESCRIPTION
**Pull request by dustindorroh**
## What changes are proposed in this pull request?
Match tensorflow auto logging to work with tensorflow object detection api. #1964 

- Added evaluate to tensorflow autologging    
- evaluate will log checkpoint files as artifacts
- log all evaluation metrics that are a number (skip serialized images)
- Changed tensorflow signature key to match tensorflow object detection api

## How is this patch tested?
Ran training on [Pets example](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_pets.md) 

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [x] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by nikste** [comment]
any update on this already?

**Comment by dustindorroh** [comment]
> any update on this already?
Well it's working for my use case. I see that the tests fail for loading tensorflow model files. Also since the output of object_detection isn't flat the serving needs some work. Still needs some testing which I don't have the time for yet. 

@nikste If you want me me to make it real pull request I will. 


image

**Comment by nikste** [comment]
would be very useful to have this as a real PR :)

**Comment by nikste** [comment]
friendly ping :)

**Comment by Qsi1984** [comment]
Hi.

Any update on this? It would be very useful to me right now ;)

Best Regards

**Comment by StphMe** [comment]
Hi,

it's working for me and would also be very useful :)

Kind regards

**Comment by zishanahmed08** [comment]
@dustindorroh Any updates on this?

**Comment by willembressers** [comment]
It's working for my project, many many thanks.

For everyone who want's to use this pull request before it's accepted in the code base, i've taken the following steps (from within a databricks notebook)

first apply the pull request as a patch
`%sh
cd /databricks/python/lib/python3.7/site-packages
wget https://patch-diff.githubusercontent.com/raw/mlflow/mlflow/pull/2396.patch
git apply 2396.patch`

Now read the contents of the original original tensorflow code.
`%sh
cat /tensorflow/models/research/object_detection/model_main.py`

Copy paste the contents of model_main.py in a new cell and add the following at the top of the cell
`%%writefile /tensorflow/models/research/object_detection/model_main_willem.py`

Add the appropriate modules
`from object_detection import model_hparams
from object_detection import model_lib

import os
import mlflow
import mlflow.tensorflow

flags.DEFINE_string(
    'model_dir', None, 'Path to output model directory '`

now add the mlflow integration, NOTE the experiment path is hardcoded, but change it to your experiment
`predict_input_fn = train_and_eval_dict['predict_input_fn']
  train_steps = train_and_eval_dict['train_steps']
  
  # specify the experiment to log to
  mlflow.set_experiment('/dev_projects/Beeldherkenning/object_detectie/object_detectie')
  
  # autolog the model
  mlflow.tensorflow.autolog(every_n_iter=1)
  
  with mlflow.start_run(run_name = os.environ['MODEL_VERSION']) as run:  

    # get the current run_id (for conveniance)
    model_id = run.info.run_id
    mlflow.set_tag("model_id", model_id)
    mlflow.set_tag("experiment_id", run.info.experiment_id)

    if FLAGS.checkpoint_dir:
      if FLAGS.eval_training_data:
        name = 'training_data'`

Now MLflow and Tensorflow are working together. In my case i needed to add a cell to tell MLFlow where the databricks host is and it's token
`import os 
import mlflow 
from databricks_cli.configure.provider import get_config 
os.environ['DATABRICKS_HOST'] = get_config().host 
os.environ['DATABRICKS_TOKEN'] = get_config().token`

And now i'm able to train the object detector WITH mlflow from within databricks, like so
`%%bash
export PYTHONPATH=$PYTHONPATH:/tensorflow/models/research:/tensorflow/models/research/slim
export PYTHONPATH=$PYTHONPATH:$REPO_LOCATION

# Train model (7.4 minutes for 1000 steps)
python /tensorflow/models/research/object_detection/model_main_willem.py \
    --logtostderr \
    --pipeline_config_path=${PIPELINE_CONFIG_PATH} \
    --model_dir=${MODEL_DIR} \
    --num_train_steps=${NUM_TRAIN_STEPS} \`

I hope this helps anyone.

**Comment by huberemanuel** [comment]
I'm trying to run this PR with the ssd_mobilenet_v2_fpnlite_320x320_coco17_tpu-8.config but none metric os logged at the end of the training. Did anyone had the same issue?

**Comment by mohammedayub44** [comment]
@dustindorroh  Great work on this. Recently started using MLFlow and Object Detection API, running into same issues. 
Any timeline when this will be merged into master ? 
Thanks !

**Comment by dustindorroh** [comment]
I’m glad this has been useful. Sorry I’ve been lazy on this. I’ll try and
find some time and get this into master.

**Comment by mohammedayub44** [comment]
@dustindorroh  Thanks. Wondering if this patch will work on TF 2.X object detection as well ? TF2.X uses `train_loop` and `eval_continuously` functions instead of estimators.

**Comment by mohammedayub44** [comment]
@dustindorroh Tested this out locally and worked as expected. 

image

Logged Model and Checkpoint - 

image

Hoping this will be accepted in the next release. 

Also, I'm using multiple models and config files. I want to log some of the model parameters from the pipeline config file. 
Is there a way to get access to them in train or evaluate function or some other workaround ?
Some of them as below - 
 - Number of Classes 
 - Type of feature extractor
 - Image Resizer height and width
 - PostProcessing
`max_number_of_boxes` and `data_augmentation_options` from train_config.

Appreciate the help !

**Comment by arifagus-itt** [comment]
> It's working for my project, many many thanks.
> 
> For everyone who want's to use this pull request before it's accepted in the code base, i've taken the following steps (from within a databricks notebook)
> 
> first apply the pull request as a patch
> 
> ```
> %sh
> cd /databricks/python/lib/python3.7/site-packages
> wget https://patch-diff.githubusercontent.com/raw/mlflow/mlflow/pull/2396.patch
> git apply 2396.patch
> ```
> 
> Now read the contents of the original original tensorflow code.
> 
> ```
> %sh
> cat /tensorflow/models/research/object_detection/model_main.py
> ```
> 
> Copy paste the contents of model_main.py in a new cell and add the following at the top of the cell
> 
> ```
> %%writefile /tensorflow/models/research/object_detection/model_main_willem.py
> ```
> 
> Add the appropriate modules
> 
> ```
> from object_detection import model_hparams
> from object_detection import model_lib
> 
> import mlflow
> import mlflow.tensorflow
> 
> flags.DEFINE_string(
>     'model_dir', None, 'Path to output model directory '
> ```
> 
> now add the mlflow integration, NOTE the experiment path is hardcoded, but change it to your experiment
> 
> ```
>   train_steps = train_and_eval_dict['train_steps']
>   
>   # specify the experiment to log to
>   mlflow.set_experiment('/dev_projects/Beeldherkenning/object_detectie/object_detectie')
>   
>   # autolog the model
>   mlflow.tensorflow.autolog(every_n_iter=1)
>   
>   with mlflow.start_run(run_name = 'object_detector') as run:  
> 
>     # get the current run_id (for conveniance)
>     mlflow.set_tag("model_id", run.info.run_id)
>     mlflow.set_tag("experiment_id", run.info.experiment_id)
> 
>     if FLAGS.checkpoint_dir:
>       if FLAGS.eval_training_data:
>         name = 'training_data'
> ```
> 
> Now MLflow and Tensorflow are working together. In my case i needed to add a cell to tell MLFlow where the databricks host is and it's token
> 
> ```
> import os 
> import mlflow 
> from databricks_cli.configure.provider import get_config 
> os.environ['DATABRICKS_HOST'] = get_config().host 
> os.environ['DATABRICKS_TOKEN'] = get_config().token
> ```
> 
> And now i'm able to train the object detector WITH mlflow from within databricks, like so
> 
> ```
> %%bash
> export PYTHONPATH=$PYTHONPATH:/tensorflow/models/research:/tensorflow/models/research/slim
> export PYTHONPATH=$PYTHONPATH:$REPO_LOCATION
> 
> python /tensorflow/models/research/object_detection/model_main_willem.py \
>     --logtostderr \
>     --pipeline_config_path=${PIPELINE_CONFIG_PATH} \
>     --model_dir=${MODEL_DIR} \
>     --num_train_steps=${NUM_TRAIN_STEPS} \
> ```
> 
> I hope this helps anyone.
>
Hi, this was really helped me. I was able to log params and metrics to mlflow. But how to log models to mlflow. I have tried some many ways but still couldn't solve it. Would really appreciate if you could help.
